ttnn/cpp/ttnn/operations/data_movement/common/kernels/common.hpp

```diff
@@ -6,9 +6,82 @@
 // It's best to copy and paste the functions in rather than include the header as code size will likely explode
 // Best to separate in to cpp/hpp at some point to avoid the code size explosion but need to figure out the linking
 // issues
+#include <stdio.h>
+#include <cstring>
+#define MASK_64      0xFFFFFFFFFFFFFFC0
+#define OFFSET_64    0x000000000000003F
+#define MASK_16      0xFFFFFFFFFFFFFFF0
+#define OFFSET_16    0x000000000000000F
 
 namespace tt::data_movement::common {
 
+#define max_packet_size 8192
+
+template <uint32_t max_transfer_size, bool only_reads>
+FORCE_INLINE void enhanced_noc_async_read(
+    const uint64_t src_noc_addr, const uint32_t dst_l1_addr, const uint32_t bytes) {
+    // If you do not know the max_transfer_size at compile time write 0 to it.
+    // only reads is true if we ONLY use noc_async_read and all calls to tt_memmove have use_read_datamover as True
+    if constexpr (((max_transfer_size < max_packet_size) && (max_transfer_size != 0)) || only_reads) {
+        noc_async_read_one_packet(src_noc_addr, dst_l1_addr, bytes);
+    } else {
+        noc_async_read(src_noc_addr, dst_l1_addr, bytes);
+    }
+}
+
+template <uint32_t max_transfer_size, bool only_writes>
+FORCE_INLINE void enhanced_noc_async_write(
+    const uint32_t src_l1_addr, const uint64_t dst_noc_addr, const uint32_t bytes) {
+    // If you do not know the max_transfer_size at compile time write 0 to it.
+    // only writes is true if we ONLY use noc_async_read and all calls to tt_memmove have use_read_datamover as False
+    if constexpr (((max_transfer_size < max_packet_size) && (max_transfer_size != 0)) || only_writes) {
+        noc_async_write_one_packet(src_l1_addr, dst_noc_addr, bytes);
+    } else {
+        noc_async_write(src_l1_addr, dst_noc_addr, bytes);
+    }
+}
+
+template <bool guaranteed_16B_alligned, bool copy_async, bool use_read_datamover, uint32_t max_transfer_size>
+FORCE_INLINE void tt_memmove(const uint32_t dst_l1_addr, const uint32_t src_l1_addr, const uint32_t bytes) {
+    //Function performs a memory copy between two l1 addresses in the local core
+    //Uses noc_async_read when possible to copy the data over
+    //Set guaranteed 16B alligned to true if the source and destination are externally guaranteed to be 16B alligned (dangerous)
```

Comment:
You should also not be hardcoding alignment values like 16 and 64 since there's no guarantee they're always that value across arches, these can be queried and passed from host as compile time args or accessed in kernels ex L1_ALIGNMENT.

Reply:
This is a bit annoying to fix as I am also using MASK and OFFSET that is hardcoded to what should be used if the shift 16 and 64 respectively and they are also used in other files as this is the common code for all data mover ops. I can't take the value as compile time arg as this is in common. However because it is in common I think updating it should a new arch use a new alignment is not very difficult.

Reply:
Come to think of it, I think the cleanest fix would be to define a DDR_ALIGNMENT at the same location L1_ALIGNMENT is defined that is set automatically at compile time depending on the ARCH type. The OPs would then use that constant value instead which is auto updated to match the architecture. However that would not be the solution for this PR

Reply:
Would it be acceptable to let this merge in as is for now and then I will post an issue where I can fix it in a cleaner way, just in the interest of getting this in as soon as possible.

Reply:
DRAM_ALIGNMENT is also already defined (is this what you meant by DDR_ALIGNMENT?). It's fine if you want to address later, I'm not blocking the pr on any of these comments but wanted to make sure this was considered, as there is currently work being done regarding what alignments are being used.

Reply:
Oh cool, good to know. Yeah will do a subsequent PR since I do want to avoid having to re-visit alignment when the next arch comes in